src/picongpu/include/plugins/ChargeConservation.tpp

```diff
@@ -0,0 +1,218 @@
+                       checkpointDirectory );
+}
+
+namespace detail
+{
+
+/**
+ * @class Div
+ * @brief divergence functor for 2D and 3D
+ *
+ * NOTE: This functor uses a Yee-cell stencil.
+ */
+template<int dim, typename ValueType>
+struct Div;
+
+template<typename ValueType>
+struct Div<DIM3, ValueType>
+{
+    typedef ValueType result_type;
+
+    template<typename Field>
+    HDINLINE ValueType operator()(Field field) const
+    {
+        const ValueType reciWidth = float_X(1.0) / cellSize.x();
+        const ValueType reciHeight = float_X(1.0) / cellSize.y();
+        const ValueType reciDepth = float_X(1.0) / cellSize.z();
+        return ((*field).x() - (*field(-1,0,0)).x()) * reciWidth +
+               ((*field).y() - (*field(0,-1,0)).y()) * reciHeight +
+               ((*field).z() - (*field(0,0,-1)).z()) * reciDepth;
+    }
+};
+
+template<typename ValueType>
+struct Div<DIM2, ValueType>
+{
+    typedef ValueType result_type;
+
+    template<typename Field>
+    HDINLINE ValueType operator()(Field field) const
+    {
+        const ValueType reciWidth = float_X(1.0) / cellSize.x();
+        const ValueType reciHeight = float_X(1.0) / cellSize.y();
+        return ((*field).x() - (*field(-1,0)).x()) * reciWidth +
+               ((*field).y() - (*field(0,-1)).y()) * reciHeight;
+    }
+};
+
+// functor for all species to calculate density
+template<typename T_SpeciesName, typename T_Area>
+struct ComputeChargeDensity
+{
+    typedef typename T_SpeciesName::type SpeciesName;
+    static const uint32_t area = T_Area::value;
+
+    HINLINE void operator()( FieldTmp* fieldTmp,
+                             const uint32_t currentStep) const
+    {
+        DataConnector &dc = Environment<>::get().DataConnector();
+
+        /* load species without copying the particle data to the host */
+        SpeciesName* speciesTmp = &(dc.getData<SpeciesName >(SpeciesName::FrameType::getName(), true));
+
+        /* run algorithm */
+        typedef typename CreateDensityOperation<SpeciesName>::type::Solver ChargeDensitySolver;
+        fieldTmp->computeValue < area, ChargeDensitySolver > (*speciesTmp, currentStep);
+        dc.releaseData(SpeciesName::FrameType::getName());
+    }
+};
+
+} // namespace detail
+
+void ChargeConservation::notify(uint32_t currentStep)
+{
+    typedef SuperCellSize BlockDim;
+
+    DataConnector &dc = Environment<>::get().DataConnector();
+
+    /* load FieldTmp without copy data to host */
+    FieldTmp* fieldTmp = &(dc.getData<FieldTmp > (FieldTmp::getName(), true));
+    /* reset density values to zero */
+    fieldTmp->getGridBuffer().getDeviceBuffer().setValue(FieldTmp::ValueType(0.0));
+
+    /* calculate and add the charge density values from all species in FieldTmp */
+    ForEach<VectorAllSpecies, picongpu::detail::ComputeChargeDensity<bmpl::_1,bmpl::int_<CORE + BORDER> >, MakeIdentifier<bmpl::_1> > computeChargeDensity;
+    computeChargeDensity(forward(fieldTmp), currentStep);
```

Comment:
Due to different treatment of borders in hdf5 and this code, this density calculation might be a source of numeric errors. 

The charge density is approximately zero due to the equal number of positively and negatively charged particles. However, here we subtract the charge density of the second species several times in parts. Thus floating point issues might cause the large difference.